db/revtree.go

```diff
@@ -44,29 +44,32 @@ type RevTree map[string]*RevInfo
 // rev IDs, with a parallel array of parent indexes. Ordering in the arrays doesn't matter.
 // So the parent of Revs[i] is Revs[Parents[i]] (unless Parents[i] == -1, which denotes a root.)
 type revTreeList struct {
-	Revs           []string          `json:"revs"`                 // The revision IDs
-	Parents        []int             `json:"parents"`              // Index of parent of each revision (-1 if root)
-	Deleted        []int             `json:"deleted,omitempty"`    // Indexes of revisions that are deletions
-	Bodies_Old     []string          `json:"bodies,omitempty"`     // JSON of each revision (legacy)
-	BodyMap        map[string]string `json:"bodymap,omitempty"`    // JSON of each revision
-	BodyKeyMap     map[string]string `json:"bodyKeyMap,omitempty"` // Keys of revision bodies stored in external documents
-	Channels       []base.Set        `json:"channels"`
-	HasAttachments []int             `json:"hasAttachments,omitempty"` // Indexes of revisions that has attachments
+	Revs           []string            `json:"revs"`                     // The revision IDs
+	Parents        []int               `json:"parents"`                  // Index of parent of each revision (-1 if root)
+	Deleted        []int               `json:"deleted,omitempty"`        // Indexes of revisions that are deletions
+	Bodies_Old     []string            `json:"bodies,omitempty"`         // Deprecated: JSON of each revision (see bodymap)
+	BodyMap        map[string]string   `json:"bodymap,omitempty"`        // JSON of each revision
+	BodyKeyMap     map[string]string   `json:"bodyKeyMap,omitempty"`     // Keys of revision bodies stored in external documents
+	Channels_Old   []base.Set          `json:"channels,omitempty"`       // Deprecated: Channels for revisions (see channelsMap)
+	ChannelsMap    map[string]base.Set `json:"channelsMap,omitempty"`    // Channels for non-winning leaf revisions (current channels stored outside of history, and don't have a use-case for storing non-leaf channels)
+	HasAttachments []int               `json:"hasAttachments,omitempty"` // Indexes of revisions that has attachments
 }
 
 func (tree RevTree) MarshalJSON() ([]byte, error) {
 	n := len(tree)
 	rep := revTreeList{
-		Revs:     make([]string, n),
-		Parents:  make([]int, n),
-		Channels: make([]base.Set, n),
+		Revs:    make([]string, n),
+		Parents: make([]int, n),
 	}
 	revIndexes := map[string]int{"": -1}
 
+	winner, _, _ := tree.winningRevision(context.TODO())
```

Comment:
Can we move this check to revtree.UnmarshalJSON when Channels_Old is present, such that we only populate info.Channels for non-winning revisions?  (i.e. it behaves like ChannelMap, and we can assume that info.Channels is only populated for non-winning leaf revisions.)

In addition to standardizing the handling for non-winning revisions, it would have the  benefit of only incurring the cost of winningRevision when Channels_Old is present.

Reply:
Should be done with latest commit